### DISCUSSION TITLE
Support for Prisma as underlying ORM

### DISCUSSION THREAD

**Discussion by Alfagun74** [ppetzold/nestjs-paginate#905] [Ideas]
TypeORM isn't really being actively maintained or enhanced anymore, and it's facing some serious bugs. Prisma, on the other hand, is gaining popularity rapidly and by now even has more Github Stars than the good ol' TypeORM. I'm considering moving my NestJS Apps to Prisma gradually. One obstacle is my reliance on NestJS Paginate Controllers, which are closely tied to TypeORM. I love this library. However, I'm wondering if it's possible to make this library work with both TypeORM and Prisma. What are your thoughts on adding Prisma support as a secondary option?

**Comment by Helveg** [comment]
Do you have any links to official statements of the authors that TypeORM won't be maintained anymore? All I could find in this regard is perhaps https://github.com/typeorm/typeorm/issues/3267?

The codebase currently has no separation from TypeORM and is tightly coupled to its query builder interface.

**Reply by Alfagun74** [reply]
While there may not be an official statement explicitly stating this, the commit and release history and the lack of willingness to fix bug-tickets that are open for years increasingly demonstrate the degradation of the ORM.

> The codebase currently has no separation from TypeORM and is tightly coupled to its query builder interface.

I think it would make sense to implement this seperation. Prisma is currently one of the most widely used TypeScript ORMs available. 

To achieve this, we could create an ORM interface or a similar abstraction layer and implement that for each ORM. This would separate nestjs-paginate from TypeORM, serving as a bridge. With this setup, switching between different underlying ORMs would be seamless.

Also this project would not die alongside TypeORM if that should be the case anytime.

Also users of other ORMs like MikroORM could implement their own interface into the library if they wanted etc.

**Reply by Helveg** [reply]
That sounds good, I'm currently spending all my available time on https://github.com/ppetzold/nestjs-paginate/pull/883

Would you be willing to spend some time in analysing the codebase to design the required interface? I think analysing the `paginate` function and `PaginateConfig` would be sufficient to get an overview of the requirements `nestjs-paginate` has for the ORM.

**Reply by Alfagun74** [reply]
I would love to, but unfortunately I'm knee-deep in running two businesses + a family, and working full-time, so I don't think I'll be able to work my way into the codebase and design anything of substance. :(

**Reply by xMase** [reply]
I need it too (typeorm have too much bug, currently virtual colums cant be diplayed if not selected explicitly and anyone accept the pr that fix the error), I can help to produce something if anyone else want to join the fork.

**Reply by fpnuseis** [reply]
It's a reasonable argument. Ultimately, this repo must support Prisma, or a new repo that supports Prisma must be created.
